build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/transport/CollectTransportVersionReferencesTask.java

```diff
@@ -0,0 +1,150 @@
+        try (var jar = new JarInputStream(Files.newInputStream(file))) {
+            ZipEntry entry;
+            while ((entry = jar.getNextEntry()) != null) {
+                String filename = entry.getName();
+                if (filename.endsWith(CLASS_EXTENSION) && filename.endsWith(MODULE_INFO) == false) {
+                    addNamesFromClass(results, jar, classname(entry.toString()));
+                }
+            }
+        }
+    }
+
+    private void addNamesFromClass(Set<TransportVersionUtils.TransportVersionReference> results, InputStream classBytes, String classname)
+        throws IOException {
+        ClassVisitor classVisitor = new ClassVisitor(Opcodes.ASM9) {
+            @Override
+            public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+                return new MethodNode(Opcodes.ASM9, access, name, descriptor, signature, exceptions) {
+                    int lineNumber = -1;
+
+                    @Override
+                    public void visitLineNumber(int line, Label start) {
+                        lineNumber = line;
+                    }
+
+                    @Override
+                    public void visitMethodInsn(int opcode, String owner, String name, String descriptor, boolean isInterface) {
+                        if (owner.equals(TRANSPORT_VERSION_SET_CLASS) && name.equals(TRANSPORT_VERSION_SET_METHOD_NAME)) {
+                            var abstractInstruction = this.instructions.getLast();
+                            String location = classname + " line " + lineNumber;
+                            if (abstractInstruction instanceof LdcInsnNode ldcInsnNode
+                                && ldcInsnNode.cst instanceof String tvName
+                                && tvName.isEmpty() == false) {
+                                results.add(new TransportVersionUtils.TransportVersionReference(tvName, location));
+                            } else {
+                                // The instruction is not a LDC with a String constant (or an empty String), which is not allowed.
+                                throw new RuntimeException(
+                                    "TransportVersion.fromName must be called with a non-empty String literal. " + "See " + location + "."
+                                );
+                            }
+                        }
+                        super.visitMethodInsn(opcode, owner, name, descriptor, isInterface);
+                    }
+                };
+            }
+        };
+        ClassReader classReader = new ClassReader(classBytes);
+        classReader.accept(classVisitor, 0);
+    }
+
+    private static String classname(String filename) {
+        return filename.substring(0, filename.length() - CLASS_EXTENSION.length()).replace('/', '.');
```

Comment:
Do we need to make this platform independent? The path separator on Windows won't be `/`, will it?